File: leetcode/medium/Q79_WordSearch.py
from typing import List


class Solution:
    visit = []
    board = []
    offset = [[-1, 0], [0, 1], [1, 0], [0, -1]]

    def exist(self, board: List[List[str]], word: str) -> bool:
        self.visit = [[False] * len(board[0]) for i in range(len(board))]
        self.board = board
        for i in range(0, len(board)):
            for j in range(0, len(board[0])):
                if board[i][j] == word[0]:
                    self.visit[i][j] = True
                    if self.find(word, j, i, 0):
                        return True
                    self.visit[i][j] = False
        return False

    def find(self, word: str, x: int, y: int, idx: int) -> bool:
        if idx == len(word) - 1:
            return True

        h = len(self.board)
        w = len(self.board[0])
        for off in self.offset:
            dx = x + off[0]
            dy = y + off[1]
            if 0 <= dx < w and 0 <= dy < h:
                if not self.visit[dy][dx] and word[idx + 1] == self.board[dy][dx]:
                    self.visit[dy][dx] = True
                    if self.find(word, dx, dy, idx + 1):
                        return True
                    self.visit[dy][dx] = False
        return False

File: src/medium/Q79_WordSearch.java
package medium;

import java.util.LinkedList;
import java.util.Queue;

public class Q79_WordSearch {
    class Solution {
        public boolean exist(char[][] board, String word) {
            char[] wordArr = word.toCharArray();
            int m = board.length;
            int n = board[0].length;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if(wordArr[0] == board[i][j]){
                        boolean[][] visited = new boolean[m][n];
                        visited[i][j] = true;
                        if(find(j, i, board, visited, wordArr, 1)) return true;
                    }
                }
            }
            return false;
        }
        public boolean find(int x, int y, char[][] board, boolean[][] visited, char[] word, int idx){
            if(word.length == idx){
                return true;
            }
            int yMax = board.length;
            int xMax = board[0].length;

            if(x - 1 >= 0 && word[idx] == board[y][x-1] && !visited[y][x-1]){
                visited[y][x-1] = true;
                if(find(x-1, y, board, visited, word, idx + 1)) return true;
                visited[y][x-1] = false;
            }
            if(y - 1 >= 0 && word[idx] == board[y-1][x] && !visited[y-1][x]){
                visited[y-1][x] = true;
                if(find(x, y-1, board, visited, word, idx + 1)) return true;
                visited[y-1][x] = false;
            }
            if(x + 1 < xMax && word[idx] == board[y][x+1] && !visited[y][x+1]){
                visited[y][x+1] = true;
                if(find(x+1, y, board, visited, word, idx + 1)) return true;
                visited[y][x+1] = false;
            }
            if(y + 1 < yMax && word[idx] == board[y+1][x] && !visited[y+1][x]){
                visited[y+1][x] = true;
                if(find(x, y+1, board, visited, word, idx + 1)) return true;
                visited[y+1][x] = false;
            }
            return false;
        }
    }
}
